### PR TITLE
Modifying https://github.com/cortexproject/cortex/pull/4422 to retry only for GET Requests

### PR DESCRIPTION
**Pull request by krishnateja325**
**What this PR does**:
- Modifying https://github.com/cortexproject/cortex/pull/4422 to retry only for GET Requests.
-  AlertManager unary APIs are failing due a single alert manager fail as distributor picks a random one to forward the request. This change only pick the replicas alert managers in case of an error.
**Which issue(s) this PR fixes**:
Fixes #4422

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/alertmanager/distributor.go [hunk 1]
```diff
@@ -254,14 +254,32 @@ func (d *Distributor) doUnary(userID string, w http.ResponseWriter, r *http.Requ
...
+
+	// Randomize the list of instances to not always query the same one.
+	rand.Shuffle(len(instances), func(i, j int) {
+		instances[i], instances[j] = instances[j], instances[i]
+	})
+
+	for i, instance := range instances {
+		resp, err := d.doRequest(ctx, instance, req)
+		// Only return error if there is no more instances to try
+		if err != nil && i == len(instances)-1 {
```
Comment (alanprot) [inline pkg/alertmanager/distributor.go:267]: hummm..

i don't think this is doing what u wanna do.

We should retry only on "GET" requests? but this condition will be `true` only on the last instance and... we will keep trying (or retrying) on all instances.. no? Am i missing something?

**Review by alanprot — COMMENTED** [review]
Instead of doing by Verb only i think we should do by Verb and Path (as we do when to decide if is a Unary of Quorum.

File: pkg/alertmanager/distributor.go (hunk 1 above)
Comment (krishnateja325) [reply pkg/alertmanager/distributor.go]: yes,  I missed a check there. Modified the logic to address this.

File: pkg/alertmanager/distributor.go [hunk 2]
```diff
@@ -254,14 +254,36 @@ func (d *Distributor) doUnary(userID string, w http.ResponseWriter, r *http.Requ
 	defer sp.Finish()
 	// Until we have a mechanism to combine the results from multiple alertmanagers,
 	// we forward the request to only only of the alertmanagers.
-	amDesc := replicationSet.Instances[rand.Intn(len(replicationSet.Instances))]
-	resp, err := d.doRequest(ctx, amDesc, req)
-	if err != nil {
-		respondFromError(err, w, logger)
-		return
-	}
+	if req.GetMethod() == "GET" {
```
Comment (danielblando) [inline pkg/alertmanager/distributor.go:257]: It seems we are just reusing the old code if not GET, which is ok but just duplicate the code

I would prefer to use the same for for both code. I think there are a few options like changing the size of the array to 1 if it is not a GET or have a retry counter which can be different for each api method.

nit: Also i would probably not check for 
`i == len(instances)-1`
i think we can keep a hold of the last error and throw it if the for finish without succeeds.

Comment (krishnateja325) [reply pkg/alertmanager/distributor.go]: Modified the logic to change the size of the array to 1 if it is not a get request.
for the last comment `keep a hold of the last error and throw it if the for loop finish without succeeds.` 
this check `if err != nil && i == len(instances)-1 {` also does the same thing right. can you let me know if I am missing something here?

**Comment by krishnateja325** [comment]
> Instead of doing by Verb only i think we should do by Verb and Path (as we do when to decide if is a Unary of Quorum.

Added a check to retry only when the request is a GET request and isUnaryReadPath

File: pkg/alertmanager/distributor_test.go [hunk 3]
```diff
@@ -196,6 +198,25 @@ func TestDistributor_DistributeRequest(t *testing.T) {
...
+			replicationFactor:  3,
+			isRead:             true,
+			expStatusCode:      http.StatusInternalServerError,
+			expectedTotalCalls: 3,
+			route:              "/status",
+		}, {
+			name:               "Read /status is sent to 3 AM when 2 are not happy",
+			numAM:              3,
+			numHappyAM:         3,
+			failedReqCount:     2,
```
Comment (qinxx108) [inline pkg/alertmanager/distributor_test.go:214]: what's the different between failedReqCount and numHappyAM?

Comment (krishnateja325) [reply pkg/alertmanager/distributor_test.go:214]: numHappyAM means number of healthy Alert Managers and failedReqCount is the number of Requests made to these AMs that are failed.

**Review by qinxx108 — APPROVED** [review]

**Review by danielblando — APPROVED** [review]

**Review by alvinlin123 — APPROVED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -2088,3 +2088,4 @@ This release has several exciting features, the most notable of them being setti
 * [FEATURE] You can specify "heap ballast" to reduce Go GC Churn #1489
 * [BUGFIX] HA Tracker no longer always makes a request to Consul/Etcd when a request is not from the active replica #1516
 * [BUGFIX] Queries are now correctly cancelled by the query-frontend #1508
+* [ENHANCEMENT] AlertManager: Retrying AlertManager Get Requests (Get Alertmanager status, Get Alertmanager Receivers) on next replica on error #4840
```
Comment (alvinlin123) [inline CHANGELOG.md:2091]: New changelog entry should be on top, notunder the 0.1.0 release.

Comment (krishnateja325) [reply CHANGELOG.md]: updated the changelog entry and placed this change under  ## master / unreleased

**Review by alanprot — APPROVED** [review]

**Pull request merged**